refactor!: Remove sailthru and braze specific modules

mikix opened the pull request:
The diff may be a little hard to read, but basically what happened was:
- Existing code had `sailthru/v1` and `braze/v1` folders. The `sailthru` folder actually had both sailthru and braze implementations, depending on config. The `braze` folder held some core braze functionality used by the braze code in `sailthru`.
- I renamed `sailthru/v1` to `email/v1`
- I moved all the braze code in `braze/v1` and `email/v1` into `email/v1/braze`
- I deleted the sailthru pieces in `email/v1` and the now-empty `braze/v1` folder
- The public API is all in `email/v1/api` and is just tiny wrappers that check config, and if braze is enabled, calls into the actual implementations defined in `email/v1/braze`. Here's the list of public functions in `api.py`:
  - send_code_assignment_nudge_email
  - send_offer_assignment_email
  - send_offer_update_email
  - send_offer_usage_email
  - did_email_bounce

See related PR on the ecommerce side to use this refactored API: https://github.com/edx/ecommerce/pull/3416

**Merge checklist:**
- [x] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production
    - `test.in` for test requirements
    - `make upgrade && make requirements` have been run to regenerate requirements
- [x] [Version](https://github.com/edx/ecommerce-worker/blob/master/setup.py) bumped

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/ecommerce-worker/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/ecommerce-worker), verify version has been pushed to [PyPI](https://pypi.org/project/edx-ecommerce-worker/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [ecommerce](https://github.com/edx/ecommerce) to upgrade dependencies (including ecommerce-worker)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in ecommerce will look for the latest version in PyPi.
    - Note: the ecommerce-worker constraint in ecommerce **must** also be bumped to the latest version in PyPi.
- [ ] Deploy `ecommerce`
- [ ] Deploy `ecomworker` on GoCD.
    - While some functions in ecommerce-worker are run via ecommerce, others are handled by a standalone AMI and must be
      released via GoCD.

MatthewPiatetsky commented on ecommerce_worker/configuration/base.py:
```diff
@@ -15,22 +15,22 @@
 
 # Backend used to store task results.
-# See http://celery.readthedocs.org/en/latest/configuration.html#celery-result-backend.
+# See http://celery.readthedocs.org/en/4.0/userguide/configuration.html#std:setting-result_backend.
```
what are these changes about?

mikix replied:
Did you try going to the first link? It's dead.

MatthewPiatetsky commented on ecommerce_worker/email/v1/utils.py:
```diff
@@ -0,0 +1,47 @@
+    try:
+        api_response = api.status().post(post_data)
+    except RequestException as e:
+        print("MIKE: exception", e)
```
are these supposed to be here?

mikix replied:
Thanks for catch btw - this is now fixed.

Dillon-Dumesnil commented on ecommerce_worker/email/v1/utils.py:
```diff
@@ -0,0 +1,47 @@
+    try:
+        api_response = api.status().post(post_data)
+    except RequestException as e:
+        print("MIKE: exception", e)
```
BRUH

Dillon-Dumesnil commented on ecommerce_worker/email/v1/utils.py:
```diff
@@ -0,0 +1,47 @@
+    except RequestException as e:
+        print("MIKE: exception", e)
+        return False
+    print("MIKE: response", api_response)
```
Part 2

Dillon-Dumesnil reviewed (approved):
Some small cleanup, but nothing looks very out of place. Assuming with coverage and all, you get a resounding sure from me

mikix commented:
I made some minor changes, FYI:
- added some new tests to improve coverage
- ran `make upgrade` rather than manually editing the requirements files, so there are quite a few changes in those folders - but that's ok, because these requirements files don't really get used - ecommerce's files are what matters. Ours only gets used for tests I believe
- updated version to `2.0.0`, for this breaking change
- dropped `MIKE` prints, whoops

inventhouse commented on ecommerce_worker/email/v1/braze/tasks.py:
```diff
@@ -0,0 +1,154 @@
+        )
+
+
+def send_code_assignment_nudge_email_via_braze(self, email, subject, email_body, sender_alias, site_code):
```
These methods appear to be very repetitive, could they (or at least their implementations) be consolidated?

mikix replied:
You mean the exception handling? I could see a decorator or something making sense to handle the common cases. But as I said above, I was trying to just focus on moving code, not actually writing any. Almost all the code here is just shuffled around.

inventhouse commented on ecommerce_worker/email/v1/tasks.py:
```diff
@@ -0,0 +1,92 @@
+    send_offer_usage_email_via_braze,
+)
+
+# pylint: disable=unused-argument
```
Might be good to note why there are unused arguments

mikix replied:
Fair - I'll add a comment, but just for clarity - I left the unused arguments because old implementations (sailthru) used them and possibly future ones might be interested in that same info (or braze might become interested), so as a matter of public API, I just left them in.

inventhouse commented on ecommerce_worker/email/v1/tests/test_api.py:
```diff
@@ -0,0 +1,23 @@
+        """
+        Test that we don't accidentally expose anything we don't expect to
+        """
+        print(getmembers(api))
```
Is this `print` leftover?  If not, why are we `print`ing?

mikix replied:
Excellent catch - just me forgetting to clean up after debugging a test.

inventhouse reviewed (approved):
Some comments/questions but mostly fine

mikix commented on ecommerce_worker/email/v1/braze/tasks.py:
```diff
@@ -0,0 +1,154 @@
+                        token_offer=offer_assignment_id,
+                        token_email=user_email,
+                    )
+                )
```
Great question, I didn't notice that code path. But this is not code I wrote, I'm just moving it around. (It's originally from sailthru/v1/tasks.py, line 225)

So I agree that looks like an oversight, but I'd like to avoid actually changing any behavior in this one.
